### ISSUE TITLE
Path B: Heckman selection coupling between classification and regression heads

### ISSUE THREAD

**Issue by Polichinel** [views-platform/views-hydranet#39]
## Sequence Position

> **This is Step 3b of 6** in the zero-inflation instability remediation sequence: **A → E → D/B → C/F**
>
> **Depends on:** #36 (Path A: Tobit loss) — must be validated first.
> **Parallel with:** #38 (Path D: Tail-aware extension) — these two address different calibration gaps and can be evaluated independently.
>
> This is a **refinement** issue, not a stability fix. Only pursue if Path A produces a measurable selection bias in active-cell predictions.

---

### What we already have

The model has two parallel output pathways per target (6 decoder heads total in `HydraBNrecurrentUnet_06_LSTM4.py`):

- **Classification head:** Predicts P(conflict > 0) — binary event probability (raw logits → sigmoid during inference at `hydranet_inference.py:254`)
- **Regression head:** Predicts E[fatalities | fatalities > 0] — continuous intensity in log1p-space (output after ReLU)

These heads are trained jointly via the multi-task loss (`mtloss.py:39-73`) with learnable homoscedastic uncertainty weights, but are otherwise **statistically independent** — the classification head's output does not inform the regression head, and vice versa.

### The Heckman selection problem

Zhang et al. (2021) identify that this dual-head architecture is structurally a **Heckman selection model** — but with a missing component. In the classical Heckman formulation:

```
Selection equation:   d* = w'γ + u       (latent propensity for conflict)
                      d = 1 if d* > 0     (binary: conflict observed)

Outcome equation:     y* = x'β + ε       (latent intensity)
                      y = y* if d = 1     (intensity observed only when conflict occurs)

Correlation:          corr(u, ε) = ρ      (selection and outcome errors are correlated)
```

The key insight: when ρ ≠ 0, the regression head is trained on a **non-random subsample** (only cells where conflict occurred). The conditional expectation is biased:

```
E[y | d=1, x] = x'β + ρ·σ_ε · λ(w'γ)
```

where λ(·) = φ(·)/Φ(·) is the **inverse Mills ratio** — a correction term that accounts for the fact that we only observe intensity for selected (conflict-active) cells.

### What this means for our model

If the classification and regression heads share latent features (which they do — shared encoder), then ρ is likely nonzero. Without the correction, the regression head's predictions for active cells may be systematically biased upward (because it's only trained on the portion of the distribution where the outcome exceeds the selection threshold).

With Path A (Tobit), this bias is partially mitigated — the Tobit loss includes a censored component for zero cells. But the Tobit model assumes the same latent variable drives both censoring and intensity. The Heckman model relaxes this by allowing two correlated but distinct latent processes (one for "will there be conflict" and one for "how much conflict").

---

## Literature Support

| Paper | Authors | Year | Key finding | Link |
|-------|---------|------|-------------|------|
| Deep Tobit Networks (DTN-II) | Zhang, Li, Song, Ning | 2021 | DTN-II architecture implements Heckman selection with deep networks. Separate selection and outcome networks connected via inverse Mills ratio and learnable ρ. Outperforms DTN-I (standard Tobit) on selection-biased data. | [ScienceDirect](https://www.sciencedirect.com/science/article/abs/pii/S0893608021003531) |
| Classical Heckman | Heckman | 1979 | Original two-step estimator for sample selection bias. Identifies that regression on a selected sample produces biased estimates unless corrected. | Econometrica 47(1) |

---

### The correction term

Add an inverse Mills ratio correction from the classification head's output to the regression head's prediction:

```python
# During training:
# 1. Get classification logits (selection equation)
class_logits = model_output.cls[:, j, :, :]  # raw logits for target j

# 2. Compute inverse Mills ratio
phi = torch.distributions.Normal(0, 1).log_prob(class_logits).exp()  # PDF
Phi = torch.distributions.Normal(0, 1).cdf(class_logits)             # CDF
lambda_imr = phi / (Phi + 1e-8)  # inverse Mills ratio

# 3. Correct regression prediction
rho_sigma = self.rho[j] * self.sigma  # learnable parameters
corrected_pred = reg_pred[:, j, :, :] + rho_sigma * lambda_imr
```

### What changes

1. **New learnable parameter ρ** (one per target, or shared). Initialized to 0 (no correction initially).
2. **Modified training loss:** The regression loss is computed on `corrected_pred` rather than raw `reg_pred`.
3. **During inference:** Apply the same correction using the classification head's output.

### What does NOT change

- Model architecture (no new heads or layers)
- Classification loss (unchanged)
- Autoregressive loop (unchanged)
- The classification and regression heads remain separate decoders — they share the encoder but have independent decoder weights

---

## Diagnostic Criteria

"Do we have a selection bias problem?" can be assessed after Path A is running:

1. **Compute ρ empirically:** Correlate classification residuals with regression residuals across all active cells. If |ρ| > 0.1, the correction matters.
2. **Conditional calibration:** Plot predicted vs actual intensity for cells where P(conflict) is near the decision boundary (e.g., P ∈ [0.3, 0.7]). Systematic overprediction in this region suggests selection bias.
3. **Compare with/without correction:** Train with ρ=0 (no correction, baseline) and ρ learned. If CRPS improves significantly, selection bias was present.

If ρ ≈ 0 empirically, this path adds complexity without benefit. **Do not implement unless the diagnostic shows a meaningful correlation.**

---

## Risks

- **MTL interference:** Adding a coupling between classification and regression heads may interfere with the existing learnable MTL loss weighting (`mtloss.py`). The MTL weights assume independent losses; the Heckman correction introduces a dependency.
- **Training instability:** ρ is a sensitive parameter. If ρ drifts during training, the correction term can amplify rather than correct. May need to clip ρ to [-0.5, 0.5] or add regularization.
- **Computational cost:** The inverse Mills ratio requires computing φ and Φ at every cell — minor cost but adds to the per-step overhead.

---

## Acceptance Criteria

- [ ] Selection bias diagnostic completed after Path A validation
- [ ] Decision made: correction needed vs ρ ≈ 0
- [ ] If correction needed: ρ parameter implemented, trained, validated
- [ ] If correction not needed: documented rationale with empirical evidence
- [ ] CRPS comparison: corrected vs uncorrected
- [ ] All existing tests pass

**Comment by Polichinel** [comment]
**PARKED** — deferred fallback (not rejected); see ZINB epic #97 / checklist #104 / C-139. Revisit only if the ZINB head fails. Rationale: loss/calibration investigate-X — deferred; ZINB is the chosen direction
